8317692: jcmd GC.heap_dump performance regression after JDK-8292818

SirYwell opened the pull request:
See the bug description for more information.

This implementation brings down the time to take a heap dump on the example application in the bug report to <2 seconds on my machine.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8317692](https://bugs.openjdk.org/browse/JDK-8317692): jcmd GC.heap_dump performance regression after JDK-8292818 (**Bug** - P4)


### Reviewers
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**) ⚠️ Review applies to [f303a227](https://git.openjdk.org/jdk/pull/16083/files/f303a2272dd2dd787a956e52507bbec73f587fff)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16083/head:pull/16083` \
`$ git checkout pull/16083`

Update a local copy of the PR: \
`$ git checkout pull/16083` \
`$ git pull https://git.openjdk.org/jdk.git pull/16083/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16083`

View PR using the GUI difftool: \
`$ git pr show -t 16083`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16083.diff">https://git.openjdk.org/jdk/pull/16083.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16083#issuecomment-1751405594)

SirYwell commented:
/label add serviceability

alexmenkov commented on src/hotspot/share/oops/fieldStreams.hpp:
```diff
@@ -179,4 +180,104 @@ class AllFieldStream : public FieldStreamBase {
+
+  void prepare() {
+    _next_klass = next_klass_with_fields();
+    // special case: the base klass has no fields. If any supertype has any fields, use that directly.
```
"base klass" sounds misleading.
I think "initial" would be clearer

alexmenkov commented:
The fix itself looks good to me.
How did you tested the change?
Looks like we don't have test coverage for the correctness of the dumped fields. Would be nice to add it.

SirYwell commented:
> The fix itself looks good to me. How did you tested the change? Looks like we don't have test coverage for the correctness of the dumped fields. Would be nice to add it.

Thanks.
I ran `hotspot_serviceability` and also manually looked into more complex heap dumps. I agree that specific tests would be better. I'll need to figure out how that can be accomplished. If you have any pointers how to get started there, please let me know.

alexmenkov commented:
> > The fix itself looks good to me. How did you tested the change? Looks like we don't have test coverage for the correctness of the dumped fields. Would be nice to add it.
> 
> Thanks. I ran `hotspot_serviceability` and also manually looked into more complex heap dumps. I agree that specific tests would be better. I'll need to figure out how that can be accomplished. If you have any pointers how to get started there, please let me know.

We have test library to parse hprof files in test/lib/jdk/test/lib/hprof
You can look at test/hotspot/jtreg/serviceability/jvmti/vthread/HeapDump/VThreadInHeapDump.java as an example of a test which generates heap dump for target application and verifies it contains expected data.

SirYwell commented:
> We have test library to parse hprof files in test/lib/jdk/test/lib/hprof You can look at test/hotspot/jtreg/serviceability/jvmti/vthread/HeapDump/VThreadInHeapDump.java as an example of a test which generates heap dump for target application and verifies it contains expected data.

Thanks, I added a test case that ensures that the instance fields are all present. This is a very basic test, but it covers super types and also makes sure the order of supertypes is correct. If you want me to add something, please let me know.

alexmenkov commented on test/hotspot/jtreg/serviceability/HeapDump/FieldsInInstanceTest.java:
```diff
@@ -0,0 +1,153 @@
+            log("Snapshot resolved.");
+
+            Iterable<JavaHeapObject> objects = snapshot.getThings()::asIterator;
+            for (JavaHeapObject heapObj : objects) {
```
Instead if iteration over all dumped objects it would be faster to:
`
       (JavaObject)snapshot.findClass(className).getInstances(false).nextElement();
`

alexmenkov commented:
> Thanks, I added a test case that ensures that the instance fields are all present. This is a very basic test, but it covers super types and also makes sure the order of supertypes is correct. If you want me to add something, please let me know.

Could you add testcases for corner cases:
no fields:
  interface I1 {
  }
  class NoFields1 {
  }
  class NoFields2 extends NoFields1 implements I1 {
  }

no parent fields:
  class NoParentFields extends NoFields1 implements I1 {
    int i1 = 1;
    int i2 = 2;
  }

only parent fields:
  class Parent1 {
    int i3 = 3;
  }
  class OnlyParentFields extends Parent1 {
  }

SirYwell commented:
> > Thanks, I added a test case that ensures that the instance fields are all present. This is a very basic test, but it covers super types and also makes sure the order of supertypes is correct. If you want me to add something, please let me know.
> 
> Could you add testcases for corner cases: no fields: interface I1 { } class NoFields1 { } class NoFields2 extends NoFields1 implements I1 { }
> 
> no parent fields: class NoParentFields extends NoFields1 implements I1 { int i1 = 1; int i2 = 2; }
> 
> only parent fields: class Parent1 { int i3 = 3; } class OnlyParentFields extends Parent1 { }

Done. I also added a test case where a class class in the "middle" class has no fields.

alexmenkov commented on test/hotspot/jtreg/serviceability/HeapDump/FieldsInInstanceTest.java:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    interface I {
+            int i = -10;
```
wrong indent

SirYwell commented:
Fixed the wrong indent. Thank you for your review. Do I need another one or can we proceed?

alexmenkov commented:
> Fixed the wrong indent. Thank you for your review. Do I need another one or can we proceed?

Hotspot changes require 2 reviewers

SirYwell commented:
/reviewers 2

dholmes-ora reviewed (commented):
This looks reasonable to me but I'd really like @fparain to review this. Thanks,

fparain reviewed (approved):
Changes look good to me.
Thank you for fixing this.

Fred

SirYwell commented:
Thank you all for your reviews.
I think this is worth a backport to jdk21u, what do you think? And if yes, is it enough to run the backport command on the commit later and create a PR from it?

/integrate

dholmes-ora commented:
/sponsor
